File: backend/src/models/Donation.js
'use strict';

const { Model, DataTypes } = require('sequelize');

module.exports = (sequelize) => {
  class Donation extends Model {
    static associate(models) {
      // Define associations here if needed
      // For now, donations are standalone
    }
  }

  Donation.init({
    id: {
      type: DataTypes.BIGINT,
      primaryKey: true,
      autoIncrement: true
    },
    stripe_payment_intent_id: {
      type: DataTypes.STRING,
      allowNull: false,
      unique: true
    },
    stripe_customer_id: {
      type: DataTypes.STRING,
      allowNull: true
    },
    amount: {
      type: DataTypes.DECIMAL(10, 2),
      allowNull: false
    },
    currency: {
      type: DataTypes.STRING(3),
      allowNull: false,
      defaultValue: 'usd'
    },
    donation_type: {
      type: DataTypes.ENUM('one-time', 'recurring'),
      allowNull: false
    },
    frequency: {
      type: DataTypes.ENUM('weekly', 'monthly', 'quarterly', 'yearly'),
      allowNull: true
    },
    payment_method: {
      type: DataTypes.ENUM('card', 'ach'),
      allowNull: false
    },
    status: {
      type: DataTypes.ENUM('pending', 'succeeded', 'failed', 'canceled'),
      allowNull: false,
      defaultValue: 'pending'
    },
    donor_first_name: {
      type: DataTypes.STRING,
      allowNull: false
    },
    donor_last_name: {
      type: DataTypes.STRING,
      allowNull: false
    },
    donor_email: {
      type: DataTypes.STRING,
      allowNull: false
    },
    donor_phone: {
      type: DataTypes.STRING,
      allowNull: true
    },
    donor_address: {
      type: DataTypes.TEXT,
      allowNull: true
    },
    donor_zip_code: {
      type: DataTypes.STRING,
      allowNull: true
    },
    metadata: {
      type: DataTypes.JSONB,
      allowNull: true
    }
  }, {
    sequelize,
    modelName: 'Donation',
    tableName: 'donations',
    timestamps: true,
    createdAt: 'created_at',
    updatedAt: 'updated_at'
  });

  return Donation;
}; 

File: backendJava/src/main/java/church/abunearegawi/backend/model/Donation.java
package church.abunearegawi.backend.model;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.Map;

@Entity
@Table(name = "donations")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Donation {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "stripe_payment_intent_id", nullable = false, unique = true)
    private String stripePaymentIntentId;

    @Column(name = "stripe_customer_id")
    private String stripeCustomerId;

    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal amount;

    @Column(nullable = false, length = 3)
    @Builder.Default
    private String currency = "usd";

    @Enumerated(EnumType.STRING)
    @Column(name = "donation_type", nullable = false)
    private DonationType donationType;

    @Enumerated(EnumType.STRING)
    private Frequency frequency;

    @Enumerated(EnumType.STRING)
    @Column(name = "payment_method", nullable = false)
    private PaymentMethod paymentMethod;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    @Builder.Default
    private Status status = Status.pending;

    @Column(name = "donor_first_name")
    private String donorFirstName;

    @Column(name = "donor_last_name")
    private String donorLastName;

    @Column(name = "donor_email")
    private String donorEmail;

    @Column(name = "donor_phone")
    private String donorPhone;

    @Column(name = "donor_address", columnDefinition = "TEXT")
    private String donorAddress;

    @Column(name = "donor_zip_code")
    private String donorZipCode;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(columnDefinition = "jsonb")
    private Map<String, Object> metadata;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    public enum DonationType {
        @Column(name = "one-time")
        ONE_TIME("one-time"),
        @Column(name = "recurring")
        RECURRING("recurring");

        private final String value;

        DonationType(String value) {
            this.value = value;
        }
    }

    public enum Frequency {
        weekly, monthly, quarterly, yearly
    }

    public enum PaymentMethod {
        card, ach
    }

    public enum Status {
        pending, succeeded, failed, canceled
    }
}
